File: node/src/events/drainer.ts
import type { Identity } from "../connection/service-bridge";
import type { EventsClient } from "../pb/servicebridge/v1/events";
import { PublishStatus } from "../pb/servicebridge/v1/events";
import type { Storage } from "../sqlite/storage";
import type { Logger } from "./publisher";

// MAX_BACKOFF_MS is the delay every attempt past the ladder reuses: transient
// failures retry forever, the ladder only caps the rate.
// @internal
const MAX_BACKOFF_MS = 600_000;
// BACKOFF_MS is the retry delay ladder for transient failures (±25% jitter
// applied).
// @internal
const BACKOFF_MS: readonly number[] = [
	1_000,
	5_000,
	30_000,
	120_000,
	MAX_BACKOFF_MS,
];

// SELECT_DUE_SQL reads the next batch of deliverable rows. Column order matches
// event_outbox_pending_order_idx, so SQLite walks the index in ORDER BY order
// and LIMIT stops the scan — no temp b-tree sort of the whole backlog.
// @internal — см. ./README.md
export const SELECT_DUE_SQL = `SELECT * FROM event_outbox
  WHERE status='pending' AND next_attempt_at_ms <= ?
  ORDER BY enqueued_at_ms ASC, id ASC
  LIMIT ?`;

// SELECT_NEXT_ATTEMPT_SQL finds when the earliest deferred row becomes due, so
// the idle wait lasts exactly that long instead of polling.
// @internal
const SELECT_NEXT_ATTEMPT_SQL = `SELECT MIN(next_attempt_at_ms) AS next_at_ms
  FROM event_outbox WHERE status='pending'`;

// RETRY_SQL re-arms a row for a later attempt.
// @internal
const RETRY_SQL = `UPDATE event_outbox
  SET status='pending', attempts=?, last_error=?, next_attempt_at_ms=?
  WHERE id=?`;

// FAIL_SQL parks a row that no retry can fix.
// @internal
const FAIL_SQL = `UPDATE event_outbox
  SET status='failed', attempts=?, last_error=?
  WHERE id=?`;

// OutboxRow is a row selected from event_outbox.
// @internal
interface OutboxRow {
	id: string;
	name: string;
	payload: Buffer;
	payload_json: Buffer;
	contract_hash: string;
	partition_key: string;
	idempotency_key: string;
	fire_and_forget: number;
	headers: string;
	occurred_at_ms: number;
	enqueued_at_ms: number;
	status: string;
	attempts: number;
	last_error: string;
	next_attempt_at_ms: number;
	x_sb_trace: string;
}

// @public — см. ./README.md
export interface DrainerDeps {
	storage: Storage;
	rpcClient: EventsClient;
	identity: () => Identity | null;
	batchSize: number;
	logger: Logger;
	clockFn?: () => number;
	sleepFn?: (ms: number, signal: AbortSignal) => Promise<void>;
	// Invoked when the runtime rejects a publish for policy reasons
	// (PUBLISH_STATUS_REJECTED_FORBIDDEN). Lets the owner surface a
	// `policy_violation` event for an async (outbox) denial that can't be
	// thrown back to the publish() caller.
	onPolicyViolation?: (v: {
		declaration: string;
		value: string;
		denySide: string;
		reason: string;
	}) => void;
}

// Drainer reads pending rows from the SQLite outbox, publishes them to the
// runtime via Events.Publish, and updates their status based on the response.
// Edge-triggered: kick() schedules an immediate wakeup from the wait state.
// @public — см. ./README.md
export class Drainer {
	private readonly deps: DrainerDeps;
	private readonly clockFn: () => number;
	private readonly sleepFn: (ms: number, signal: AbortSignal) => Promise<void>;

	private running = false;
	private pendingKick = false;
	private wakeResolve: (() => void) | null = null;
	private loopDone: Promise<void> = Promise.resolve();

	constructor(deps: DrainerDeps) {
		this.deps = deps;
		this.clockFn = deps.clockFn ?? (() => Date.now());
		this.sleepFn = deps.sleepFn ?? defaultSleep;
	}

	// start begins the background drain loop.
	start(): void {
		if (this.running) return;
		this.running = true;
		this.loopDone = this.loop();
	}

	// kick wakes up the drainer immediately if it is waiting.
	// Edge-triggered: if already running an iteration, marks pendingKick
	// so the next wait is also skipped.
	kick(): void {
		this.pendingKick = true;
		if (this.wakeResolve) {
			const resolve = this.wakeResolve;
			this.wakeResolve = null;
			resolve();
		}
	}

	// stop signals the loop to exit and waits for the current iteration to finish.
	async stop(): Promise<void> {
		this.running = false;
		this.kick(); // wake up from any wait
		await this.loopDone;
	}

	private async loop(): Promise<void> {
		while (this.running) {
			const now = this.clockFn();
			const { storage, rpcClient, identity, batchSize } = this.deps;

			// Select pending rows whose next_attempt_at_ms is due.
			const rows = storage
				.prepare(SELECT_DUE_SQL)
				.all(now, batchSize) as OutboxRow[];

			if (rows.length === 0) {
				if (!this.running) break;
				await this.waitForWork();
				continue;
			}

			// Claim: mark rows as inflight so a crash-restart resets them.
			const ids = rows.map((r) => r.id);
			const placeholders = ids.map(() => "?").join(",");
			storage
				.prepare(
					`UPDATE event_outbox SET status='inflight' WHERE id IN (${placeholders})`,
				)
				.run(...ids);

			// Build and send PublishRequest.
			const ident = identity();
			const events = rows.map((r) => ({
				id: r.id,
				name: r.name,
				payload: r.payload,
				payloadJson: r.payload_json ?? Buffer.alloc(0),
				contractHash: r.contract_hash,
				partitionKey: r.partition_key,
				idempotencyKey: r.idempotency_key,
				fireAndForget: false,
				headers: (() => {
					try {
						return JSON.parse(r.headers) as Record<string, string>;
					} catch {
						return {};
					}
				})(),
				occurredAtUnixMs: r.occurred_at_ms,
				xSbTrace: r.x_sb_trace ?? "",
			}));

			let results: {
				eventId: string;
				status: PublishStatus;
				message: string;
			}[] = [];
			let netError: Error | null = null;

			try {
				const response = await new Promise<{
					results: {
						eventId: string;
						status: PublishStatus;
						message: string;
					}[];
				}>((resolve, reject) => {
					rpcClient.publish(
						{
							publisherServiceId: ident?.serviceId ?? "",
							publisherInstanceId: ident?.instanceId ?? "",
							events,
						},
						(err, res) => {
							if (err) reject(err);
							else
								resolve(
									res ?? {
										results: [],
									},
								);
						},
					);
				});
				results = response.results;
			} catch (err) {
				netError = err instanceof Error ? err : new Error(String(err));
			}

			if (netError) {
				// The runtime is unreachable. Transport failure says nothing about
				// the event itself, so it never burns a terminal budget: rows are
				// re-armed with the (saturating) backoff ladder and retried for as
				// long as the outbox lives. That is the whole point of the outbox.
				const message = netError.message;
				storage.transaction(() => {
					for (const row of rows) {
						this.rearm(row, message);
					}
				});
				this.deps.logger.error(
					`drainer: publish of ${rows.length} event(s) failed, will retry — ${message}`,
				);
				// Continue loop — will respect backoff via next_attempt_at_ms filter.
				continue;
			}

			// Process per-event results. One transaction for the whole batch: a
			// per-row autocommit would append a WAL frame and a commit record per
			// event.
			const resultMap = new Map(results.map((r) => [r.eventId, r]));
			const doneIds: string[] = [];
			const violations: {
				declaration: string;
				value: string;
				denySide: string;
				reason: string;
			}[] = [];
			let transientCount = 0;

			storage.transaction(() => {
				for (const row of rows) {
					const entry = resultMap.get(row.id);
					const status =
						entry?.status ?? PublishStatus.PUBLISH_STATUS_UNSPECIFIED;

					switch (status) {
						case PublishStatus.PUBLISH_STATUS_ACCEPTED:
						case PublishStatus.PUBLISH_STATUS_REJECTED_DUPLICATE:
							// Success or idempotent duplicate — delete from outbox.
							doneIds.push(row.id);
							break;

						case PublishStatus.PUBLISH_STATUS_REJECTED_INVALID_NAME:
							// Terminal: the name will not become valid on a retry.
							storage
								.prepare(FAIL_SQL)
								.run(row.attempts + 1, `rejected:${status}`, row.id);
							break;

						case PublishStatus.PUBLISH_STATUS_REJECTED_FORBIDDEN: {
							// Policy denial (event.publish gate). Terminal — retrying
							// won't help; mark failed and surface a policy_violation so
							// the async denial is observable (publish() is fire-and-forget
							// into the outbox, so it can't be thrown back to the caller).
							const reason = entry?.message || "event.publish denied by policy";
							storage
								.prepare(FAIL_SQL)
								.run(row.attempts + 1, `forbidden:${reason}`, row.id);
							violations.push({
								declaration: "event.publish",
								value: row.name,
								denySide: "self_egress",
								reason,
							});
							break;
						}

						default:
							// Unspecified means the runtime could not say anything about
							// this event — transient, same infinite-retry treatment as a
							// transport failure.
							transientCount++;
							this.rearm(row, "unspecified");
							break;
					}
				}

				if (doneIds.length > 0) {
					const donePlaceholders = doneIds.map(() => "?").join(",");
					storage
						.prepare(
							`DELETE FROM event_outbox WHERE id IN (${donePlaceholders})`,
						)
						.run(...doneIds);
					storage.adjustOutboxRowCount(-doneIds.length);
				}
			});

			// Observers run after commit: onPolicyViolation publishes an event of
			// its own, which would re-enter storage.transaction().
			if (transientCount > 0) {
				this.deps.logger.warn(
					`drainer: UNSPECIFIED status for ${transientCount} event(s) — treating as transient`,
				);
			}
			for (const violation of violations) {
				this.deps.logger.warn(
					`drainer: event ${violation.value} rejected — ${violation.reason}`,
				);
				this.deps.onPolicyViolation?.(violation);
			}

			// If kick was requested during this iteration, loop immediately.
			if (this.pendingKick) {
				this.pendingKick = false;
			}
		}
	}

	// rearm schedules the row for another attempt on the backoff ladder. The
	// last rung repeats forever — the runtime being down must not consume a
	// delivery budget.
	private rearm(row: OutboxRow, lastError: string): void {
		const attempts = row.attempts + 1;
		const base = BACKOFF_MS[attempts - 1] ?? MAX_BACKOFF_MS;
		const jitter = Math.floor(base * 0.25 * (Math.random() * 2 - 1));
		this.deps.storage
			.prepare(RETRY_SQL)
			.run(attempts, lastError, this.clockFn() + base + jitter, row.id);
	}

	// waitForWork blocks until kick() fires or the earliest deferred row is due.
	// With nothing pending it waits for kick() alone — an idle SDK schedules no
	// timers at all.
	private async waitForWork(): Promise<void> {
		if (this.pendingKick) {
			this.pendingKick = false;
			return;
		}

		const next = this.deps.storage.prepare(SELECT_NEXT_ATTEMPT_SQL).get() as {
			next_at_ms: number | null;
		} | null;
		const nextAtMs = next?.next_at_ms ?? null;
		if (nextAtMs !== null && nextAtMs <= this.clockFn()) return;

		const controller = new AbortController();
		await new Promise<void>((resolve) => {
			this.wakeResolve = resolve;
			if (nextAtMs !== null) {
				void this.sleepFn(nextAtMs - this.clockFn(), controller.signal).then(
					resolve,
				);
			}
		});
		// Cancels the pending timer when kick() won the race.
		controller.abort();
		this.wakeResolve = null;
		this.pendingKick = false;
	}
}

// defaultSleep resolves after ms, or immediately when the signal aborts,
// clearing the timer either way.
// @internal
function defaultSleep(ms: number, signal: AbortSignal): Promise<void> {
	return new Promise<void>((resolve) => {
		const timer = setTimeout(resolve, ms);
		signal.addEventListener(
			"abort",
			() => {
				clearTimeout(timer);
				resolve();
			},
			{ once: true },
		);
	});
}

File: go/internal/events/drainer.go
package events

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/service-bridge/sdk/go/internal/outbox"
	pb "github.com/service-bridge/sdk/go/internal/pb/servicebridge/v1"
	"github.com/service-bridge/sdk/go/internal/stream"
)

// DefaultBatchSize is how many rows one drain iteration claims.
const DefaultBatchSize = 100

// DefaultPublishTimeout bounds one Events.Publish call.
const DefaultPublishTimeout = 30 * time.Second

// RetryLadder is the delay between drain attempts for one row. It is a rate
// limiter, not a budget: the last rung repeats forever. A ladder that expired
// would mark every buffered event failed after a few minutes of downtime — and
// the publish that buffered them already returned success and promised
// durability.
func RetryLadder() stream.Backoff {
	return stream.NewBackoff(
		stream.WithLadder(
			1*time.Second,
			5*time.Second,
			30*time.Second,
			2*time.Minute,
			10*time.Minute,
		),
		stream.WithJitterRatio(0.25),
	)
}

// PolicyViolation reports an event the runtime refused on policy grounds. The
// publish that buffered it returned long ago, so this is the only way the
// denial reaches the owner.
type PolicyViolation struct {
	EventID   string
	EventName string
	Reason    string
}

// DrainerConfig wires the drain loop. See ./README.md.
type DrainerConfig struct {
	Storage  *outbox.Storage
	Publish  PublishFunc
	Identity func() Identity
	// BatchSize defaults to DefaultBatchSize.
	BatchSize int
	// Backoff pins the retry ladder. Zero value falls back to RetryLadder().
	Backoff stream.Backoff
	// PublishTimeout defaults to DefaultPublishTimeout.
	PublishTimeout time.Duration
	// OnPolicyViolation surfaces a terminal policy denial to the owner. It runs
	// on the drain goroutine after the batch is committed.
	OnPolicyViolation func(PolicyViolation)
	// OnError reports storage failures. The loop keeps running.
	OnError func(error)
	// Now returns unix-ms; defaults to the wall clock.
	Now func() int64
	// Sleep waits out d or returns early when ctx ends. Defaults to a timer.
	Sleep  func(ctx context.Context, d time.Duration)
	Logger *slog.Logger
}

// Drainer moves buffered events to the runtime. One goroutine owns the loop;
// Stop cancels it and waits.
type Drainer struct {
	cfg DrainerConfig

	// kick is edge-triggered and coalescing. Buffered so a wake raised while an
	// iteration runs survives until the next wait instead of being lost.
	kick chan struct{}

	wg sync.WaitGroup

	mu      sync.Mutex
	started bool
	cancel  context.CancelFunc
}

// NewDrainer validates the config and fills its defaults.
func NewDrainer(cfg DrainerConfig) (*Drainer, error) {
	if cfg.Storage == nil {
		return nil, fmt.Errorf("events: new drainer: missing Storage: %w", ErrInvalidConfig)
	}
	if cfg.Publish == nil {
		return nil, fmt.Errorf("events: new drainer: missing Publish: %w", ErrInvalidConfig)
	}
	if cfg.Identity == nil {
		return nil, fmt.Errorf("events: new drainer: missing Identity: %w", ErrInvalidConfig)
	}
	if cfg.BatchSize <= 0 {
		cfg.BatchSize = DefaultBatchSize
	}
	if len(cfg.Backoff.Rungs()) == 0 {
		cfg.Backoff = RetryLadder()
	}
	if cfg.PublishTimeout <= 0 {
		cfg.PublishTimeout = DefaultPublishTimeout
	}
	if cfg.Now == nil {
		cfg.Now = func() int64 { return time.Now().UnixMilli() }
	}
	if cfg.Sleep == nil {
		cfg.Sleep = sleepCtx
	}
	if cfg.Logger == nil {
		cfg.Logger = slog.Default()
	}
	return &Drainer{cfg: cfg, kick: make(chan struct{}, 1)}, nil
}

// Start launches the drain goroutine. Single-use: Stop is terminal.
func (d *Drainer) Start(ctx context.Context) error {
	d.mu.Lock()
	if d.started {
		d.mu.Unlock()
		return fmt.Errorf("events: start drainer: %w", ErrAlreadyStarted)
	}
	runCtx, cancel := context.WithCancel(ctx)
	d.started = true
	d.cancel = cancel
	d.mu.Unlock()

	d.wg.Add(1)
	go d.run(runCtx)
	return nil
}

// Stop cancels the loop and waits for it. Idempotent.
func (d *Drainer) Stop() {
	d.mu.Lock()
	cancel := d.cancel
	d.cancel = nil
	d.mu.Unlock()

	if cancel != nil {
		cancel()
	}
	d.wg.Wait()
}

// Kick wakes an idle drain. Non-blocking and coalescing.
func (d *Drainer) Kick() {
	select {
	case d.kick <- struct{}{}:
	default:
	}
}

func (d *Drainer) run(ctx context.Context) {
	defer d.wg.Done()

	errAttempt := 0
	for {
		if ctx.Err() != nil {
			return
		}
		n, err := d.drainOnce(ctx)
		switch {
		case err != nil:
			if ctx.Err() != nil {
				return
			}
			d.reportError(err)
			// A storage failure must not spin the loop: back off like any other
			// failure and try again.
			d.cfg.Sleep(ctx, d.cfg.Backoff.Delay(errAttempt))
			errAttempt++
		case n > 0:
			errAttempt = 0
		default:
			errAttempt = 0
			d.waitForWork(ctx)
		}
	}
}

// drainOnce claims one batch, hands it to the runtime and applies the outcome.
// It returns how many rows it claimed.
func (d *Drainer) drainOnce(ctx context.Context) (int, error) {
	recs, err := d.cfg.Storage.ClaimDue(ctx, d.cfg.Now(), d.cfg.BatchSize)
	if err != nil {
		return 0, fmt.Errorf("events: drain: claim: %w", err)
	}
	if len(recs) == 0 {
		return 0, nil
	}

	resp, sendErr := d.send(ctx, recs)
	result, violations := d.classify(recs, resp, sendErr)

	if err := d.cfg.Storage.Complete(ctx, result); err != nil {
		// The batch stays inflight until the next Open resets it. Nothing is
		// lost, but nothing moves either, so this is loud.
		return 0, fmt.Errorf("events: drain: complete: %w", err)
	}

	for _, v := range violations {
		d.cfg.Logger.Warn("events: drain: event rejected by policy",
			"event_id", v.EventID, "event_name", v.EventName, "reason", v.Reason)
		if d.cfg.OnPolicyViolation != nil {
			d.cfg.OnPolicyViolation(v)
		}
	}
	return len(recs), nil
}

func (d *Drainer) send(ctx context.Context, recs []outbox.Record) (*pb.PublishResponse, error) {
	ident := d.cfg.Identity()
	events := make([]*pb.EventEnvelope, len(recs))
	for i, r := range recs {
		events[i] = &pb.EventEnvelope{
			Id:               r.ID,
			Name:             r.Name,
			Payload:          r.Payload,
			PayloadJson:      r.PayloadJSON,
			ContractHash:     r.ContractHash,
			PartitionKey:     r.PartitionKey,
			IdempotencyKey:   r.IdempotencyKey,
			FireAndForget:    r.FireAndForget,
			Headers:          r.Headers,
			OccurredAtUnixMs: r.OccurredAtMs,
			XSbTrace:         r.Trace,
		}
	}
	callCtx, cancel := context.WithTimeout(ctx, d.cfg.PublishTimeout)
	defer cancel()

	resp, err := d.cfg.Publish(callCtx, &pb.PublishRequest{
		PublisherServiceId:  ident.ServiceID,
		PublisherInstanceId: ident.InstanceID,
		Events:              events,
	})
	if err != nil {
		return nil, fmt.Errorf("events: drain: publish %d event(s): %w", len(recs), err)
	}
	return resp, nil
}

// classify turns one runtime answer into the row updates it implies.
//
// A transport failure says nothing about any single event, so it never parks a
// row: every row is re-armed on the saturating ladder and retried for as long
// as the outbox lives. Only a per-envelope verdict the runtime will repeat —
// a name it rejects, a policy that denies the publish — is terminal.
func (d *Drainer) classify(recs []outbox.Record, resp *pb.PublishResponse, sendErr error) (outbox.Result, []PolicyViolation) {
	var (
		result     outbox.Result
		violations []PolicyViolation
	)

	if sendErr != nil {
		d.cfg.Logger.Warn("events: drain: batch failed, retrying",
			"events", len(recs), "err", sendErr)
		for _, r := range recs {
			result.Retry = append(result.Retry, d.rearm(r, sendErr.Error()))
		}
		return result, nil
	}

	byID := make(map[string]*pb.PublishStatusEntry, len(resp.GetResults()))
	for _, entry := range resp.GetResults() {
		byID[entry.GetEventId()] = entry
	}

	transient := 0
	for _, r := range recs {
		entry := byID[r.ID]
		switch entry.GetStatus() {
		case pb.PublishStatus_PUBLISH_STATUS_ACCEPTED,
			pb.PublishStatus_PUBLISH_STATUS_REJECTED_DUPLICATE:
			result.Done = append(result.Done, r.ID)

		case pb.PublishStatus_PUBLISH_STATUS_REJECTED_INVALID_NAME:
			result.Failed = append(result.Failed, outbox.Failure{
				ID:        r.ID,
				Attempts:  r.Attempts + 1,
				LastError: "invalid name: " + entry.GetMessage(),
			})

		case pb.PublishStatus_PUBLISH_STATUS_REJECTED_FORBIDDEN:
			reason := entry.GetMessage()
			if reason == "" {
				reason = "event.publish denied by policy"
			}
			result.Failed = append(result.Failed, outbox.Failure{
				ID:        r.ID,
				Attempts:  r.Attempts + 1,
				LastError: "forbidden: " + reason,
			})
			violations = append(violations, PolicyViolation{
				EventID:   r.ID,
				EventName: r.Name,
				Reason:    reason,
			})

		default:
			// Unspecified, or no verdict at all for this id: the runtime said
			// nothing about the event, which is the transport case again.
			transient++
			result.Retry = append(result.Retry, d.rearm(r, "unspecified status"))
		}
	}
	if transient > 0 {
		d.cfg.Logger.Warn("events: drain: no verdict, retrying", "events", transient)
	}
	return result, violations
}

// rearm schedules one row for a later attempt. attempts only picks the ladder
// rung — it is never compared against a limit.
func (d *Drainer) rearm(r outbox.Record, lastError string) outbox.Retry {
	attempts := r.Attempts + 1
	delay := d.cfg.Backoff.Delay(int(r.Attempts))
	return outbox.Retry{
		ID:              r.ID,
		Attempts:        attempts,
		LastError:       lastError,
		NextAttemptAtMs: d.cfg.Now() + delay.Milliseconds(),
	}
}

// waitForWork parks until a publish kicks the loop or the earliest deferred row
// comes due. With nothing pending it waits on the kick alone — an idle SDK arms
// no timer.
func (d *Drainer) waitForWork(ctx context.Context) {
	select {
	case <-d.kick:
		// A publish landed while the iteration ran. Re-query at once.
		return
	default:
	}

	dueAt, ok, err := d.cfg.Storage.NextDueAt(ctx)
	if err != nil {
		if ctx.Err() != nil {
			return
		}
		d.reportError(fmt.Errorf("events: drain: next due: %w", err))
		d.waitKick(ctx, d.cfg.Backoff.Delay(0))
		return
	}
	if !ok {
		d.waitKick(ctx, 0)
		return
	}
	wait := dueAt - d.cfg.Now()
	if wait <= 0 {
		return
	}
	d.waitKick(ctx, time.Duration(wait)*time.Millisecond)
}

// waitKick blocks until the kick fires, ctx ends, or d elapses. A non-positive
// d means no timer at all.
func (d *Drainer) waitKick(ctx context.Context, wait time.Duration) {
	if wait <= 0 {
		select {
		case <-d.kick:
		case <-ctx.Done():
		}
		return
	}
	t := time.NewTimer(wait)
	defer t.Stop()
	select {
	case <-d.kick:
	case <-t.C:
	case <-ctx.Done():
	}
}

func (d *Drainer) reportError(err error) {
	d.cfg.Logger.Error("events: drain", "err", err)
	if d.cfg.OnError != nil {
		d.cfg.OnError(err)
	}
}

func sleepCtx(ctx context.Context, wait time.Duration) {
	if wait <= 0 {
		return
	}
	t := time.NewTimer(wait)
	defer t.Stop()
	select {
	case <-t.C:
	case <-ctx.Done():
	}
}
